ledger/src/ui/src/onboard.c

```diff
@@ -0,0 +1,160 @@
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ */
+
+#include <string.h>
+
+#include "apdu.h"
+#include "defs.h"
+#include "err.h"
+#include "os.h"
+#include "onboard.h"
+
+// Global onboarding flag
+extern const unsigned char *N_onboarded_ui[1];
+
+/*
+ * Reset the given onboard context
+ *
+ * @arg[out] onboard_ctx onboard context
+ */
+void reset_onboard_ctx(onboard_t *onboard_ctx) {
+    explicit_bzero(onboard_ctx, sizeof(onboard_t));
+}
+
+/*
+ * Implement the RSK WIPE command.
+ *
+ * Wipes and onboards the device.
+ *
+ * @arg[in]  pin_ctx     pin context
+ * @arg[out] onboard_ctx onboard context
+ * @ret                  number of transmited bytes to the host
+ */
+unsigned int wipe_device(onboard_t *onboard_ctx, pin_t *pin_ctx) {
+    volatile unsigned char onboarded_flag = 0;
+
+    // Reset the onboarding flag to mark onboarding
+    // hasn't been done just in case something fails
+    nvm_write((void *)PIC(N_onboarded_ui),
+              (void *)&onboarded_flag,
+              sizeof(onboarded_flag));
+
+#ifndef DEBUG_BUILD
+    if (!is_pin_valid(pin_ctx)) {
+        THROW(ERR_INVALID_PIN);
+    }
+#endif
+
+    // Wipe device
+    os_global_pin_invalidate();
+    os_perso_wipe();
+    // Generate 32 bytes of random with onboard rng
+    cx_rng((unsigned char *)onboard_ctx->client_seed, HASHSIZE);
+    // XOR with host-generated 32 bytes random
+    for (int i = 0; i < HASHSIZE; i++) {
+        onboard_ctx->client_seed[i] ^= onboard_ctx->words_buffer[i];
+    }
+    // The seed is now in client_seed, generate the mnemonic
+    os_memset(onboard_ctx->words_buffer, 0, sizeof(onboard_ctx->words_buffer));
+    onboard_ctx->words_buffer_length =
+        bolos_ux_mnemonic_from_data((unsigned char *)onboard_ctx->client_seed,
+                                    SEEDSIZE,
+                                    (unsigned char *)onboard_ctx->words_buffer,
+                                    sizeof(onboard_ctx->words_buffer));
+    // Clear the seed
+    explicit_bzero(onboard_ctx->client_seed, sizeof(onboard_ctx->client_seed));
+
+    // Set seed from mnemonic
+    os_perso_derive_and_set_seed(0,
+                                 NULL,
+                                 0,
+                                 NULL,
+                                 0,
+                                 onboard_ctx->words_buffer,
+                                 strlen(onboard_ctx->words_buffer));
+
+    // Clear the mnemonic
+    explicit_bzero(onboard_ctx->words_buffer,
+                   sizeof(onboard_ctx->words_buffer));
+
+    // Set PIN
+    os_perso_set_pin(0, GET_PIN(pin_ctx), GET_PIN_LENGTH(pin_ctx));
+    // Finalize onboarding
+    os_perso_finalize();
+    os_global_pin_invalidate();
+    unsigned char output_index = CMDPOS;
+    SET_APDU_AT(output_index++, 2);
+    SET_APDU_AT(output_index++,
+                os_global_pin_check(GET_PIN(pin_ctx), GET_PIN_LENGTH(pin_ctx)));
+
```

Comment:
I think we need to clear the pin after validating it.

Reply:
I just saw that now, sorry. And I also saw that we also clear the pin manually on every operation in that way. I think it's fine for now, but we should probably turn that into a `clear_pin`/`zero_pin` function eventually and call it within every module as part of its functionality. Thoughts?

Reply:
We'll just keep the clearing of the pin within the calling code since, as @italo-sampaio mentioned, it is indeed clearer to see which operations preseve the pin and which don't in that way.